File: Code Python/03. Compression/03.2 Compression with FFT - Fast Fourier Transformation/ImageCompressionFFT2.py
from scipy.linalg import dft
from matplotlib.image import imread
import matplotlib.pyplot as plt
import numpy as np

Image = imread('03. Compression/03.2 Compression with FFT - Fast Fourier Transformation/Einstein.jpg')
Image = np.double(Image)

# Surface plot
columnindices = np.arange(0, 2945)
columnindices[::-1].sort()
X, Y = np.meshgrid(np.arange(1, np.shape(Image)[1]+1), np.arange(1, np.shape(Image)[0]+1))
fig0 = plt.figure(figsize=(10, 10))
sp = fig0.add_subplot(111, projection='3d')
mappable = sp.plot_surface(X, Y, Image[:, columnindices], cmap='jet')
sp.view_init(65, 100)
fig0.colorbar(mappable)
plt.show()

# Plotting Discrete Fourier Transformation Matrix
F = dft(3668)
fig1 = plt.figure(figsize=(20, 20))
plt.imshow(np.real(F), cmap='gray')
plt.title('DFT matrix')

# Fourier Transormation
FourImage = np.fft.ifft2(Image)

# Reducing Data
# Percentages to which we reduce our data
perc = [0.99, 0.05, 0.01, 0.001]

# Fourier Coefficients sorted for finding % of largest
FourOrder = np.abs(FourImage.flatten())
FourOrder[::-1].sort()

fig2 = plt.figure(figsize=(20, 10))

# Looping percentages
for i in range(0, len(perc)):
    limit = FourOrder[int(np.round(perc[i]*len(FourOrder)))]
    ind = abs(FourImage) > limit
    RedFourier = FourImage*ind
    RedImage = np.rot90(np.rot90(np.fft.ifft2(RedFourier)))
    plt.subplot(2, 4, i+1)
    plt.imshow(np.abs(RedImage), cmap='gray')
    plt.axis('off')
    plt.title('Coefficients reduced to ' + str(perc[i-1]*100) + '%')
    plt.subplot(2, 4, i+5)
    Frame = RedImage[1000:1500, 1000:1700]
    plt.imshow(np.abs(Frame), cmap='gray')
    plt.axis('off')
    plt.title('Coefficients reduced to ' + str(perc[i-1]*100) + '%')

plt.show()

# Edge Detection
modes = [50, 500, 1000, 2000]
fig3 = plt.figure(figsize=(20, 10))

for i in range(0, len(modes)):
    limit = FourOrder[modes[i]]
    ind = abs(FourImage) < limit
    RedFourier = FourImage*ind
    RedImage = np.rot90(np.rot90(np.fft.ifft2(RedFourier)))
    plt.subplot(2, 4, i+1)
    plt.imshow(np.abs(RedImage), cmap='gray')
    plt.axis('off')
    plt.title('Coefficients reduced by ' + str(modes[i]) + ' frequencies')
    plt.subplot(2, 4, i+5)
    RedFourierlog = np.log(np.abs(np.fft.fftshift(RedFourier))+1)
    plt.imshow(np.abs(RedFourierlog), cmap='gray')
    plt.axis('off')
    plt.title('Coefficients reduced by ' + str(modes[i]) + ' frequencies')

plt.show()

File: Code Matlab/03. Compression/03.2 Compression with FFT - Fast Fourier Transformation/ImageCompressionFFT2.m
clc,clear,close all
%% Import Picture
Image = imread('Einstein.png');
Image = double(Image);

%% Plotting Surface
fig1 = figure('Name','Surface Plot');
sp1 = subplot(1,2,1);
colormap(sp1 ,gray)
imshow(uint8(Image))
title('Image')
sp2 = subplot(1,2,2);
surf(Image(:,sort(1:2945, 'descend')),'EdgeColor', 'interp')
colormap(sp2 ,jet)
title('Image as Surface')
colorbar
view([-165,79])

%% Plotting Discrete Fourier Transformation Matrix

F = dftmtx(3668);
figure('Name','DFT Matrix')
hold on
title('DFT Matrix of size 3668 \times 3668')
imshow(real(F))


%% Fourier Transormation
FourImage = fft2(Image);


%% Reducing Data
% Percentages to which we reduce our data
perc = [0.99 0.05 0.01 0.001];

% Fourier Coefficients sorted for finding % of largest
FourOrder = sort(abs(FourImage(:)),'descend');


figure('Name','Reduced Data')
hold on

% Looping percentages
for i = 1:length(perc)
    limit = FourOrder(round(perc(i)*size(FourOrder,1)));
    ind = abs(FourImage)>limit;
    RedFourier = FourImage.*ind;
    RedImage = ifft2(RedFourier);
    subplot(2,4,i)
    imshow(uint8(RedImage));
    title(['Coefficients reduced to ',num2str(perc(i)*100),'%']);
    subplot(2,4,i+4)
    Frame = RedImage(1000:1500,1000:1700);
    imshow(uint8(Frame));
    title(['Coefficients reduced to ',num2str(perc(i)*100),'%']);
end


%% Edge Detecting

modes = [50 500 1000 2000];

figure('Name','Edge Detection')

for i = 1:length(modes)
    limit = FourOrder(modes(i));
    ind = abs(FourImage)<limit;
    RedFourier = FourImage.*ind;
    RedImage = ifft2(RedFourier);
    subplot(2,4,i)
    imshow(uint8(RedImage))
    title(['Coefficients reduced by ',num2str(modes(i)),' frequencies']);
    subplot(2,4,i+4)
    RedFourierlog = log(abs(fftshift(RedFourier))+1);
    imshow(mat2gray(RedFourierlog))
    title(['Coefficients reduced by ',num2str(modes(i)),' frequencies']);
end
